### PR TITLE
Refining unpersist style and test distributed data source before converting into DataSet

### PR DESCRIPTION
**Pull request by Litchilitchy**

### CODE REVIEW

**Comment by Litchilitchy** [comment]
@zhichao-li would this one be better? BTW I am also testing if we could skip the `isDistributed` check

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 1]
```diff
@@ -391,9 +391,13 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
...
    * @param dataSet Target DataSet to release
    */
-  def releaseDataSet(dataSet: DataSet[MiniBatch[T]]): Unit = {
-    dataSet.toDistributed().unpersist()
-    dataSet.toDistributed().originRDD().unpersist()
+  def releaseDataSets(dataSets: Array[DataSet[MiniBatch[T]]]): Unit = {
+    for (ds <- dataSets) {
+      if (ds != null) {
+        ds.toDistributed().unpersist()
+        ds.toDistributed().originRDD().unpersist()
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: filter and foreach can be apply here, but it's ok for not depending on those.

**Comment by zhichao-li** [comment]
Choose a better title for this PR?

**Comment by zhichao-li** [comment]
is there any `fit` method that need to be updated as well.

**Comment by Litchilitchy** [comment]
Not sure, would check after test is finished

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 2]
```diff
@@ -391,9 +391,13 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
    *
    * @param dataSet Target DataSet to release
    */
-  def releaseDataSet(dataSet: DataSet[MiniBatch[T]]): Unit = {
-    dataSet.toDistributed().unpersist()
-    dataSet.toDistributed().originRDD().unpersist()
+  def releaseDataSets(dataSets: Array[DataSet[MiniBatch[T]]]): Unit = {
+    for (ds <- dataSets) {
+      if (ds != null) {
+        ds.toDistributed().unpersist()
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: Check if `isDistributed` here?  Actually, both `Local` or `Distributed` DataSet should implement the `unpersist` method then we don't need to write such a dummy code here.

Comment (Litchilitchy) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: use `isInstanceOf[DistributedDataSet]`? 

I checked the code and there seems to be 5 categories of handles. `RDD, Local & DistributedImageSet, Local & DistributedTextSet`, the distributed one all called `Dataset.rdd`, seems this check would work.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 3]
```diff
@@ -391,9 +391,13 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
    *
    * @param dataSet Target DataSet to release
    */
-  def releaseDataSet(dataSet: DataSet[MiniBatch[T]]): Unit = {
-    dataSet.toDistributed().unpersist()
-    dataSet.toDistributed().originRDD().unpersist()
+  def releaseDataSets(dataSets: Array[DataSet[MiniBatch[T]]]): Unit = {
+    for (ds <- dataSets) {
+      if (ds != null && ds.isInstanceOf[DistributedDataSet[T]]) {
```
Comment (Litchilitchy) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: I think this should work..

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 4]
```diff
@@ -416,10 +420,7 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
     val valData = toDataSet(validationData, batchSize, featurePaddingParam, labelPaddingParam)
     this.fit(trainData, nbEpoch, valData)
 
-    releaseDataSet(trainData)
-    if (valData != null) {
-      releaseDataSet(valData)
-    }
+    releaseDataSets(Array(trainData, valData))
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: why do we release the dataset unconditionally here?

Comment (Litchilitchy) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: this check has been moved into `releaseDataSet` method

**Pull request merged**